dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/ArgumentTypesResolver.java

```diff
@@ -0,0 +1,200 @@
+        } else {
+            return Optional.of(ArgumentTypes.customType(typeName));
+        }
+    }
+
+    private boolean isPrimitive(TypeName in) {
+        return PLAIN_SER_DE_TYPES.containsKey(in.box());
+    }
+
+    private String planSerDeMethodName(TypeName in) {
+        String typeName = PLAIN_SER_DE_TYPES.get(in.box());
+        return Preconditions.checkNotNull(typeName, "Unknown type");
+    }
+
+    private boolean isRawRequestBody(TypeMirror in) {
+        return types.isSameType(in, requestBodyType);
+    }
+
+    private Optional<OptionalType> getOptionalType(Element paramContext, TypeMirror typeName) {
+        // "They tryin'a make me use a visitor, but I say: NO, NO NO!"
+        if (!(typeName instanceof DeclaredType)) {
+            return Optional.empty();
+        }
+        DeclaredType declaredType = (DeclaredType) typeName;
+        if (types.isSameType(declaredType, optionalIntType)) {
+            return Optional.of(ImmutableOptionalType.builder()
+                    .isPresentMethodName("isPresent")
+                    .valueGetMethodName("getAsInt")
+                    .underlyingType(integerArgumentType)
+                    .build());
+        }
+
+        if (declaredType.getTypeArguments().size() != 1) {
+            return Optional.empty();
+        }
+
+        TypeMirror innerType = Iterables.getOnlyElement(declaredType.getTypeArguments());
+        DeclaredType genericOptional = types.getDeclaredType(genericOptionalType, innerType);
+
+        if (types.isSameType(genericOptional, declaredType)) {
+            return getArgumentTypeImpl(paramContext, innerType).map(argumentType -> ImmutableOptionalType.builder()
+                    .isPresentMethodName("isPresent")
+                    .valueGetMethodName("get")
+                    .underlyingType(argumentType)
+                    .build());
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Checks if type has method {@code public String valueOf()}.
+     */
+    private boolean isCustomAndHasValueOfMethod(TypeMirror typeMirror) {
+        Element element = types.asElement(typeMirror);
+        // Unclear if this is idiomatic?
+        if (!(element instanceof TypeElement)) {
+            return false;
+        }
+        TypeElement typeElement = (TypeElement) element;
+        return MoreElements.getAllMethods(typeElement, types, elements).stream().anyMatch(this::isValueOfMethod);
+    }
+
+    private boolean isValueOfMethod(ExecutableElement executableElement) {
+        Visibility visibility = Visibility.ofElement(executableElement);
+        Set<Modifier> modifiers = executableElement.getModifiers();
+        // TODO(12345): Actually give the user good feedback on what exactly they need to add to make this type comply.
+        return executableElement.getSimpleName().toString().equals(VALUE_OF_METHOD_NAME)
+                && executableElement.getParameters().isEmpty()
```

Comment:
Shouldn't this check for size 1 of type String?

Reply:
I think I see, this is backwards -- `valueOf` is used to produce an object from a string, I think we may want toString here, or to require a custom type that can handle serialization.

Reply:
I opted for instance method? so no args, return type String?

Reply:
Oh I see what you mean, you're objecting to the method name. How about asString?

Reply:
I definitely don't think we should really use #toString, that feels too footgunny; https://stackoverflow.com/questions/29207562/how-to-use-custom-type-for-pathparam is some prior art.

Reply:
#asString feels like the least of ceremony, you can just wrap types you don't own (if you really want to do that) and impl the method. It's definitely magic, so maybe all annotations could get an Optional<Class<? extends ParamConverter>>, but that is defo more work

Reply:
Not a hug fan of wrapping types, especially with magic instance methods. An arg-wrapper type could be better, but the client API would feel dirty to use, even if it was easy to write. I'd much rather use `Class<? extends ParamConverter>` for unknown types with validation at compile time.

Reply:
I think it's reasonable to exclude custom plain parameter handling entirely for this PR given the potential complexity.